From dbd5b8ce82f930a0348c2a0ba6ca30957a895eab Mon Sep 17 00:00:00 2001
From: Paul Lemire <paul.lemire@kdab.com>
Date: Mon, 22 Nov 2021 10:43:34 +0100
Subject: [PATCH 14/17] OpenGL: Fix content not updated when using OnDemand and
 switching textures

When changing a Parameter whose value is a texture, it takes Qt3D potentially
two frames to process the scene fully before it is able to draw the updated
content.

This patch makes it so that if we know we have the flags ParametersDirty for
frame N, we request to reprocess the scene once more on frame N+1.

Change-Id: Ia635e324434f92f8423a68b72d99350d96830de6
Reviewed-by: Sean Harmer <sean.harmer@kdab.com>
(cherry picked from commit 672da697b147f039effbfd352b20f3f1891f31c6)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/plugins/renderers/opengl/renderer/renderer.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/plugins/renderers/opengl/renderer/renderer.cpp b/src/plugins/renderers/opengl/renderer/renderer.cpp
index 5c55d3bb9..7b1fbb92d 100644
--- a/src/plugins/renderers/opengl/renderer/renderer.cpp
+++ b/src/plugins/renderers/opengl/renderer/renderer.cpp
@@ -1872,6 +1872,7 @@ QVector<Qt3DCore::QAspectJobPtr> Renderer::renderBinJobs()
     // Remove previous dependencies
     m_cleanupJob->removeDependency(QWeakPointer<QAspectJob>());
 
+    const bool dirtyParametersForCurrentFrame = m_dirtyBits.marked & AbstractRenderer::ParameterDirty;
     const BackendNodeDirtySet dirtyBitsForFrame = m_dirtyBits.marked | m_dirtyBits.remaining;
     m_dirtyBits.marked = {};
     m_dirtyBits.remaining = {};
@@ -1995,6 +1996,10 @@ QVector<Qt3DCore::QAspectJobPtr> Renderer::renderBinJobs()
 
     m_dirtyBits.remaining = dirtyBitsForFrame & notCleared;
 
+    // Dirty Parameters might need 2 frames to react if the parameter references a texture
+    if (dirtyParametersForCurrentFrame)
+        m_dirtyBits.remaining |= AbstractRenderer::ParameterDirty;
+
     return renderBinJobs;
 }
 
-- 
2.36.1

